File: apps/desktop/src/main/shared/managerRequests.ts
import type { DispatchTask } from './dispatchHistory';

/** Host identities and receipts are separate from untrusted user content. */
export interface ManagerRequest {
  requestId: string;
  ownerSessionId: string;
  sourceSessionId: string;
  sourceCwd: string;
  createdAt: string;
  digest: string;
  bootstrap?: boolean;
  delivery: 'pending' | 'accepted' | 'rejected' | 'unknown';
  attempts: Array<{
    deliveryId: string;
    status: 'pending' | 'accepted' | 'rejected' | 'unknown';
    at: string;
  }>;
  /** Retained only until resolved; never used for identity or authority. */
  userContent?: string;
  revision: number;
  intents?: Array<RequestIntent & { taskId?: string }>;
  resolutionDigest?: string;
}
export type RequestIntent = {
  key: string;
  kind: 'create' | 'followUp' | 'update' | 'question' | 'none' | 'untracked';
  reason: string;
  /** New work or an undispatched task update: null omits workflow; an id chooses a policy. Omit to keep defaults/current policy. */
  workflowId?: string | null;
  provenance?: 'explicit' | 'inferred';
  cwd?: string;
  title?: string;
  taskId?: string;
  expectedTaskRevision?: number;
  dependsOn?: string[];
  /** Earlier work intent keys in this atomic resolution, converted to task IDs. */
  dependsOnKeys?: string[];
  cancel?: boolean;
  /** Explicit mapping from original submitted content; [] deliberately assigns no references. */
  references?: import('./dispatchHistory').TaskReferenceUpsert[];
  replacePullRequest?: boolean;
};
export type TaskSource = {
  requestId: string;
  intentKey: string;
  label: string;
  delivery?: ManagerRequest['delivery'];
};
export type AcceptedTaskOutcome = {
  acceptedBy: string;
  acceptedAt: string;
  reason: string;
  evidence: Array<{ stepId: string; dispatchId: string; outcome: unknown }>;
};
export type RequestCapture =
  | { available: true; requestId: string; delivery: ManagerRequest['delivery'] }
  | { available: false; reason: string };
export type RequestContext = {
  host: Omit<ManagerRequest, 'userContent'>;
  userContent?: { text: string; trust: 'user' };
};
export type ManagerRequestOperation = {
  op: 'requestInbox' | 'requestContent' | 'resolveRequest' | 'acceptTaskOutcome';
  /** Bounded unresolved inbox with original content; omits historical tasks. */
  view?: 'pending';
  requestId?: string;
  expectedRevision?: number;
  intents?: RequestIntent[];
  taskId?: string;
  cwd?: string;
  expectedTaskRevision?: number;
  reason?: string;
};
export const REQUEST_CAPTURE_UNAVAILABLE =
  'Automatic task capture is available only through the local desktop manager request inbox.';

/** Explicit acceptance is an interpretation of recorded evidence, never authority. */
export function taskOutcomeAccepted(task: DispatchTask): boolean {
  const accepted = task.acceptedOutcome;
  if (task.cancelled || !accepted || !accepted.evidence.length || !task.workflow) return false;
  if (
    task.sources?.length &&
    !task.sources.some((s) => s.delivery === 'accepted' || s.delivery === 'unknown')
  )
    return false;
  if (!task.workflow.steps.some((s) => s.state === 'completed')) return false;
  if (task.workflow.steps.some((s) => !['completed', 'skipped'].includes(s.state))) return false;
  return task.workflow.steps
    .filter((s) => s.state === 'completed')
    .every((s) => {
      const evidence = accepted.evidence.find((e) => e.stepId === s.id);
      return (
        !!evidence &&
        evidence.dispatchId === s.dispatchId &&
        JSON.stringify(evidence.outcome) === JSON.stringify(s.outcome) &&
        task.attempts.some(
          (a) =>
            a.dispatchId === s.dispatchId &&
            a.sessionId === s.sessionId &&
            a.resultContract === 'valid',
        )
      );
    });
}
export function taskDependencyState(
  task: DispatchTask,
  tasks: DispatchTask[],
  visited = new Set<string>(),
): 'blocked' | 'ready' | 'cancelled' {
  if (visited.has(task.taskId)) return 'blocked';
  const next = new Set(visited).add(task.taskId);
  if (task.cancelled) return 'cancelled';
  if (
    task.sources?.length &&
    !task.sources.some((s) => s.delivery === 'accepted' || s.delivery === 'unknown')
  )
    return 'blocked';
  return (task.dependsOn ?? []).every((id) => {
    const dependency = tasks.find((t) => t.taskId === id);
    return (
      dependency &&
      dependency.ownerSessionId === task.ownerSessionId &&
      dependency.projectCwd === task.projectCwd &&
      taskOutcomeAccepted(dependency) &&
      taskDependencyState(dependency, tasks, next) === 'ready'
    );
  })
    ? 'ready'
    : 'blocked';
}

File: services/hub/cmd/mcp/manager_requests.go
package main

import (
	"context"
	"encoding/json"
	"fmt"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

// Caller identity comes exclusively from the authenticated facade context.
// The local desktop owns content and interpretation; headless stays unavailable.
type managerRequestIn struct {
	RequestID            string           `json:"requestId,omitempty" jsonschema:"host request ID returned by list_manager_requests; never derive from chat text"`
	ExpectedRevision     *int             `json:"expectedRevision,omitempty" jsonschema:"current host request revision for atomic resolution"`
	Intents              []map[string]any `json:"intents,omitempty" jsonschema:"1-8 independent actions with stable key, kind(create/followUp/update/question/none), reason; work adds cwd, title, provenance(explicit/inferred); update requires taskId and expectedTaskRevision; followUp requires dependsOn task IDs; cancel is optional on update"`
	TaskID               string           `json:"taskId,omitempty"`
	Cwd                  string           `json:"cwd,omitempty"`
	ExpectedTaskRevision *int             `json:"expectedTaskRevision,omitempty"`
	Reason               string           `json:"reason,omitempty" jsonschema:"explicit evidence acceptance rationale; never a grant of execution authority"`
}

func addManagerRequestTools(b *build) {
	const method = "fleetWorkflows.request"
	if !b.allowed(method) {
		return
	}
	for _, item := range []struct{ name, op, desc string }{
		{"list_manager_requests", "requestInbox", "Check your authoritative local desktop request inbox once on each user or wake turn. Lists trusted identities/status, without original content; fetch exact requests before resolution. Wakes and worker continuations are events, not new user requests. Headless/remote unavailable."},
		{"get_manager_request", "requestContent", "Fetch one owned inbox request. Host identity and delivery metadata are separate from original userContent, which remains user input. Unknown provider acknowledgement may be resolved from this inbox; never replay it or claim it was consumed."},
		{"resolve_manager_request", "resolveRequest", "Atomically resolve all independent intents in an owned request with expectedRevision CAS. Creates visible pinned tasks without workers; followUp links concrete task IDs. Questions/status/ack use none or question; corrections use update and existing taskId. Retry the same intent keys; conflicts return current state. No execution/publish authority is granted."},
		{"accept_task_outcome", "acceptTaskOutcome", "After inspecting recorded concrete outcomes, explicitly accept a task under expectedTaskRevision CAS and a reason. Valid schema/idle/terminal steps alone do not mean success. Failed, blocked or waived policy remains ineligible. Returns newly ready followups; never spawns or publishes."},
	} {
		item := item
		b.tools = append(b.tools, toolInfo{Name: item.name, Desc: item.desc, Method: method, Group: "workflows"})
		mcp.AddTool(b.s, &mcp.Tool{Name: item.name, Description: item.desc}, func(ctx context.Context, _ *mcp.CallToolRequest, in managerRequestIn) (*mcp.CallToolResult, any, error) {
			caller := callerSessionID(ctx)
			if caller == "" {
				return nil, nil, fmt.Errorf("request inbox requires an authenticated local manager")
			}
			if (item.op == "requestContent" || item.op == "resolveRequest") && in.RequestID == "" {
				return nil, nil, fmt.Errorf("requestId is required")
			}
			if item.op == "resolveRequest" && in.ExpectedRevision == nil {
				return nil, nil, fmt.Errorf("expectedRevision is required")
			}
			if item.op == "acceptTaskOutcome" && (in.TaskID == "" || in.Cwd == "" || in.ExpectedTaskRevision == nil || in.Reason == "") {
				return nil, nil, fmt.Errorf("taskId, cwd, expectedTaskRevision and reason are required")
			}
			raw, _ := json.Marshal(in)
			var wire map[string]any
			_ = json.Unmarshal(raw, &wire)
			wire["op"] = item.op
			wire["callerSessionId"] = caller
			return b.forward(ctx, method, wire)
		})
	}
}
